docs/available-filters/envelope-encryption/envelope-encryption.adoc

```diff
@@ -0,0 +1,210 @@
+
+On the produce path, the filter employs a DEK reuse strategy. This means that records sent by a single connection to
+the same topic will be encrypted using the same DEK (until a time-out or encryption operations limit is reached, whichever
+occurs first).  This prevents key exhaustion and prevents excessive interactions with the KMS.
+
+On the consume path, the filter employs an LRU strategy to keep recently encountered DEKs decrypted in memory. This
+prevents excessive interactions with the KMS.
+
+=== Role of the KMS
+
+The Key Management Service (KMS) is the *secure repository* for Key Encryption Keys (KEKs). The key material of the KEK
+*never* leaves the confines of the KMS.  The KMS exposes cryptographic functions that operate against the KEKs stored
+within it.
+
+The KMS is also the *source* of Data Encryption Keys (DEKs).  When the KMS generates a DEK for a given KEK, it returns
+the DEK (which is securely generated random data), together with the encrypted DEK (which is the same data, encrypted
+using the KEK).  Note that the KMS does not *store* the DEKs - DEKs are stored encrypted as part of the Kafka record held
+by the broker.
+
+The filter uses the services of the KMS to generate DEKs, and decrypt encrypted DEKs.
+
+The KMS must be available at runtime. If the KMS is unavailable, it will become impossible to produce or consume
+records through the filter until the KMS service is restored.
+
+The filter currently has a single KMS integration with https://www.hashicorp.com/[HashiCorp Vault&#174;].  More KMS
+integrations are planned. It is also possible for a user to implement their own KMS integration.  The implementation
+must implement the {kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[KMS public API] and make the
+implementation available on the classpath with the service loader mechanism.
+
+IMPORTANT: It is recommended to use a KMS in a highly available configuration.
+
+=== Key rotation
+
+Key rotation is the practice of periodically replacing cryptographic keys with new ones.  Using key rotation is
+considered cryptographic best-practice.
+
+The filter allows for the rotation of KEKs within the KMS. When a KEK is rotated, the new key material will be applied
+to newly produced records. Existing records (which were encrypted with older versions of the KEK) remain decryptable
+as long as the previous KEK version remains present in the KMS.
+
+WARNING: If the previous KEK version is removed from the KMS, the records encrypted with that key version will become
+un-consumable (that is, the fetch will fail). In this case, the consumer offset must be advanced beyond those records.
+
+=== What exactly gets encrypted?
+
+The filter currently encrypts only *record values*.  Record keys, headers, timestamps are not encrypted.
+
+Null record values produced are passed through unencrypted.  This serves to allow tombstoning.
```

Comment:
I know what your referring too but I think it assumes too much knowledge of Kafka internals for our intended audience. I think we should either explain what tombstones are and how Kafka uses and represents them while explaining that we still support them or just ignore it. As a user I would just expect the filter to support compacted topics unless it explicitly said it didn't do so.

Thoughts?

Reply:
Maybe just adding a reference link to kafka doc: https://kafka.apache.org/documentation/#design_compactionbasics

Reply:
Could you explain where we address this comment?